File: leetcode/14D Algo2/Two pointer/844. Backspace String Compare.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<list>
#include<stack>
#include<unordered_set> 
#include<unordered_map> 
#include<algorithm>
using namespace std;

/*
Given two strings s and t, return true if they are equal 
when both are typed into empty text editors. '#' means a backspace character.

Note that after backspacing an empty text, the text will continue empty.

Input: s = "ab##", t = "c#d#"
Output: true
Explanation: Both s and t become "".
*/

class Solution1 {
public:
    bool backspaceCompare(string s, string t) {
        vector<char> v1; vector<char> v2;
        return buildString(s) == buildString(t);
    }

    stack<char> buildString(string s){
        stack<char> BS;
        for(int i = 0; i < s.size(); ++i){
            if(s[i] != '#'){
                BS.push(s[i]);
            }
            else{
                if(BS.empty()) continue;
                BS.pop();
            }
        }
        return BS;
    }
};

File: leetcode/14D Algo2/Two pointer/844. Backspace String Compare.java
import java.util.*;

/*
Input: s = "ab#c", t = "ad#c"
Output: true
Explanation: Both s and t become "ac".

Input: s = "a#c", t = "b"
Output: false
Explanation: s becomes "c" while t becomes "b".
*/

class Solution {
    public boolean backspaceCompare(String s, String t) {
        int n1 = s.length(), n2 = t.length();
        int n = Math.max(n1, n2);
        StringBuilder s1 = new StringBuilder();
        StringBuilder t1 = new StringBuilder();
        int i = 0;
        while(i < n){
            if(i < n1){
                if(s.charAt(i) == '#' && s1.length() != 0){
                    s1.deleteCharAt(s1.length() - 1);
                }
                else if(s.charAt(i) != '#'){
                    s1.append(s.charAt(i));
                }
            }
            if(i < n2){
                if(s.charAt(i) == '#' && t1.length() != 0){
                    t1.deleteCharAt(t1.length() - 1);
                }
                else if(s.charAt(i) != '#'){
                    t1.append(s.charAt(i));
                }
            }
            ++i;
        }
        return s1.toString().equals(t1.toString());
    }
}